cleverhans/attacks.py

```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
         return True
 
 
+class CarliniWagnerL2(Attack):
+    """
+    This attack was originally proposed by Carlini and Wagner. It is an
+    iterative attack that finds adversarial examples on many defenses that
+    are robust to other attacks.
+    Paper link: https://arxiv.org/abs/1608.04644
+
+    """
+    def __init__(self, model, back='tf', sess=None):
+        super(CarliniWagnerL2, self).__init__(model, back, sess)
+
+        if self.back == 'th':
+            raise NotImplementedError('Theano version not implemented.')
+
+        import tensorflow as tf
+        self.feedable_kwargs = {'y': tf.float32}
+
+        self.structural_kwargs = ['nb_classes',
+                                  'batch_size', 'confidence',
+                                  'targeted', 'learning_rate',
+                                  'binary_search_steps', 'max_iterations',
+                                  'abort_early', 'initial_const',
+                                  'clip_min', 'clip_max']
+
+    def generate(self, x, **kwargs):
+        """
+        Generate adversarial samples and return them in a Numpy array.
+
+        :param x: (required) A Numpy array with the original inputs.
+        :param y: (optional) A Numpy array with the labels that we either
+                  should target (if targeted=True) or avoid (if
+                  target=False). If None, use labels the classifier assigns.
+        :param nb_classes: The number of classes the model has.
+        :param confidence: Confidence of adversarial examples: higher produces
+                           examples that are farther away, but more strongly
+                           classified as adversarial.
+        :param batch_size: Number of attacks to run simultaneously.
+        :param targeted: True if we should perform a targetted attack, False
+                         otherwise.
+        :param learning_rate: The learning rate for the attack algorithm.
```

Comment:
I don't have a better word, do you? I use learning_rate because that's what it's used for. step_size isn't entirely accurate since adam does adaptive step sizes for a given learning rate.